src/scenic/syntax/scenic.gram

```diff
@@ -737,6 +741,71 @@ scenic_class_property_attribute: &&(
     | "dynamic" { s.Dynamic(LOCATIONS) }
 )
 
+
+# Behavior
+# --------
+
+scenic_behavior_def:
+    | "behavior" a=NAME b=['(' z=[params] ')' { z }] &&':' c=scenic_behavior_def_block {
+        s.BehaviorDef(
+            a.string,
+            args=b or self.make_arguments(None, [], None, [], None),
+            docstring=c[0],
+            header=c[1],
+            body=c[2],
+            LOCATIONS,
+        )
+     }
+
+scenic_behavior_def_block:
+    # if only preconditions and/or invariants are specified
+    | NEWLINE INDENT a=[x=STRING NEWLINE { x.string }] b=scenic_behavior_header DEDENT { (a, b, [ast.Pass()]) }
+    | NEWLINE INDENT a=[x=STRING NEWLINE { x.string }] b=[scenic_behavior_header] c=scenic_behavior_statements DEDENT { (a, b or [], c) }
+    | invalid_block
+
+scenic_behavior_statements[list]: a=scenic_behavior_statement+ { list(itertools.chain.from_iterable(a)) }
+
+# statements available inside behavior (normal statements + dynamic statements - precondition/invariant)
+scenic_behavior_statement[list]:
+    | scenic_invalid_behavior_statement
```

Comment:
Because `precondition: ...` and `invariant: ...` can be parsed as typed variable initialization, use this rule to catch any precondition and invariant definition after any other statements and raise syntax error.